### ISSUE TITLE
Add ensime java mode to java layer?

### ISSUE THREAD

**Issue by brcolow** [syl20bnr/spacemacs#5708]
Currently it seems that the java layer has only eclim support - but there is an alternative - namely: ensime (which has a java mode "enjime").

More details here: http://ensime.github.io/editors/emacs/

I am not proficient enough with LISP, Emacs, or Spacemacs to try and implement this myself, but am opening it as an issue as I believe it would be a good addition to the java layer (and, more honestly, would love to try it myself!).

Thanks.

**Comment by biocyberman** [comment]
This will be very nice. I am using ensime-based scala-mode and like it very much.

**Comment by thang1thang2** [comment]
Funny that you mention this. I just went through the trouble of doing this around yesterday (I haven't made a layer, I just shoehorned the code into the `.spacemacs` file). I'll put up what I did to get it working here in case that's helpful.

Changed `dotspacemacs-additional-packages` to include `'(ensime, sbt-mode)`

This was placed in my `dotspacemacs/user-config()` function

```elisp
  (setq exec-path (append exec-path '("/usr/local/bin")))
  (setq exec-path (append exec-path '("/usr/local/sbin")))
  (setenv "PATH" (shell-command-to-string "/bin/bash -c 'echo -n $PATH'"))

  ;; this obviously isn't required for ensime but it is for eclim. I probably don't need these two lines
  (setq eclim-eclipse-dirs "/usr/lib/eclipse"
        eclim-executable "/usr/lib/eclipse/eclim")

  (use-package ensime
    :commands ensime ensime-mode)
  (add-hook 'java-mode-hook 'ensime-mode)
  (add-hook 'scala-mode-hook 'ensime-mode)

  (use-package sbt-mode
    :commands sbt-start sbt-command
    :config
    ;; WORKAROUND: https://github.com/ensime/emacs-sbt-mode/issues/31
    ;; allows using SPACE when in the minibuffer
    (substitute-key-definition
     'minibuffer-complete-word
     'self-insert-command
     minibuffer-local-completion-map))
```

Everything after this point is just installing sbt onto your computer, and setting up stuff outside of emacs. To finish the actual *layer* off would require key binding. I would suggest a variable for the java package such as `java-backend` with the parameters either `ecim` or `ensime` and then enable each setup as necessary.

Eclim and ensime are not very compatible with each other and they will fight each other in annoying ways. Luckily since each one has different functions for everything there's no direct conflict but since keybinds are probably going to be reused for the same purpose, they should be bound depending on the backend variable set.

99% of the setup for ensime java can be used for ensime scala and vice versa. Ensime was originally designed for scala and it is only recently that it even works at all in any meaningful way with java. I might suggest looking into stealing from the scala layer and/or upgrading it in parallel with this one.

**Comment by liujoey** [comment]
for now as a workaround I disabled java layer, only added scala layer, and added these lines to my user configure for java support. It's working perfectly for a pure Java project with SBT.

```
  (setq exec-path (append exec-path '("/usr/local/bin")))
  (add-hook 'java-mode-hook 'scala/configure-ensime)
  (add-hook 'java-mode-hook 'scala/maybe-start-ensime)
```

**Comment by mahinshaw** [comment]
I updated my version of the scala layer last night to optionally start ensime when in java mode.  It's not part of the java layer, but I am happy to make a PR.  Thought I would get some consensus on whether or not it belongs with the scala layer or not.  I think it works nicely since the scala layer is mostly about ensime.

**Comment by rationalis** [comment]
I think it'd make more sense to have a variable for the java layer that selects between ensime and eclim (and java layer with ensime should not automatically enable scala layer). I'm not sure which it should default to, but I'd lean towards eclim, just because it's the one that's currently supported.

**Comment by mahinshaw** [comment]
The only issue I have with adding it to the java layer is the massive amount of code duplication.

**Comment by rationalis** [comment]
Having it under the java layer is just for the benefit of the user, I don't think there should be code duplication. I'm not too familiar with spacemacs yet, so I'm not sure if this possible (or if it would be even messier) -- but I was thinking that enabling the java layer with a variable set to use ensime, would enable the scala layer, configure it for java usage, and disable it for scala unless the actual scala layer is enabled.

**Comment by sandric** [comment]
Hm, strange, I did everything like @thang1thang2 described and looks like ensime is working but parses my `.java` files like `scala` ones - for example on line with:
```java
import java.awt.SystemColor;
```
gives me
```
ERROR: illegal start of type : line 3
```
in `ensime-compilation-results` buffer and highlights with red in file's buffer.

MB my `build.sbt` written wrong - I know literally nothing about scala, can anyone who figured everything out show me his .sbt file maybe, here's mine:
```
organization := "your.group.id"
name := "Test"
version := "1.0-SNAPSHOT"
```

**Comment by worldwar2008** [comment]
In the mode of ensime, why i can't import my own java package?
There is an warn "The object is not the member of my own java package"

**Comment by hedning** [comment]
I've started refactoring ensime out of the scala layer into it's own layer which both scala and java can depend ([branch](https://github.com/hedning/spacemacs/tree/ensime)). That seems like the most reasonable way to make the cross dependency work. So far I've just put it under `layers/+lang/ensime`, but it might be more appropriate under `+misc` or `+tools`.

**Comment by TheBB** [comment]
Sounds exciting!

**Comment by peng051410** [comment]
@thang1thang2 Can you provide a complete configuration?Thank you!

**Comment by beta1440** [comment]
@brcolow it looks like this issue can be closed
